executor/adapter.go

```diff
@@ -1254,16 +1255,27 @@ func (a *ExecStmt) GetTextToLog() string {
 func (a *ExecStmt) observeStmtBeginForTopSQL() {
 	if vars := a.Ctx.GetSessionVars(); variable.TopSQLEnabled() && vars.StmtStats != nil {
 		sqlDigest, planDigest := a.getSQLPlanDigest()
-		vars.StmtStats.OnExecutionBegin(sqlDigest, planDigest)
+		vars.StmtExecCtx = &stmtstats.StatementExecutionContext{
+			SQLDigest:  sqlDigest,
+			PlanDigest: planDigest,
+		}
+		vars.StmtStats.OnExecutionBegin(vars.StmtExecCtx)
 		// This is a special logic prepared for TiKV's SQLExecCount.
 		vars.StmtCtx.KvExecCounter = vars.StmtStats.CreateKvExecCounter(sqlDigest, planDigest)
 	}
 }
 
 func (a *ExecStmt) observeStmtFinishedForTopSQL() {
 	if vars := a.Ctx.GetSessionVars(); variable.TopSQLEnabled() && vars.StmtStats != nil {
-		sqlDigest, planDigest := a.getSQLPlanDigest()
-		vars.StmtStats.OnExecutionFinished(sqlDigest, planDigest)
+		if vars.StmtExecCtx == nil {
+			sqlDigest, planDigest := a.getSQLPlanDigest()
+			vars.StmtExecCtx = &stmtstats.StatementExecutionContext{
+				SQLDigest:  sqlDigest,
+				PlanDigest: planDigest,
+			}
+		}
+		vars.StmtExecCtx.ExecDuration = time.Since(vars.StartTime) + vars.DurationParse
+		vars.StmtStats.OnExecutionFinished(vars.StmtExecCtx)
```

Comment:
Should we pass `StartTime` and `DurationParse` in `StatementExecutionContext `? And then calculate `ExecDuration` in the implementation of `OnExecutionFinished` instead of passing the result here.

Reply:
I think `OnExecutionFinished` shouldn't know how to calculate the `ExecDuration`, so I prefer just calculate in here.

Reply:
We have done this, ExecCount is calculated inside `stmtstats`:
```go
func (s *StatementStats) OnExecutionBegin(ctx *StatementExecutionContext) {
	if ctx == nil {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	item := s.GetOrCreateStatementStatsItem(ctx.SQLDigest, ctx.PlanDigest)

	item.ExecCount++
	// Count more data here.
}
```

So this is possible:
```go
func (s *StatementStats) OnExecutionFinished(ctx *StatementExecutionContext) {
	if ctx == nil {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	item := s.GetOrCreateStatementStatsItem(ctx.SQLDigest, ctx.PlanDigest)

	item.SumExecNanoDuration += time.Since(ctx.StartTime) + ctx.DurationParse
	// Count more data here.
}
```